File: Solutions/C#/0053-MaximumSubarray/MaximumSubarray.cs
using System;

namespace MaximumSubarray
{
    public class Solution
    {
        public int MaxSubArray(int[] nums)
        {
            int[] dp = new int[nums.Length]; 

            dp[0] = nums[0];

            int max = dp[0], tmp;
            for (int i = 1; i < nums.Length; i++)
            {
                tmp = dp[i - 1] + nums[i];
                dp[i] = (tmp > nums[i] ? tmp : nums[i]);

                if (dp[i] > max)
                {
                    max = dp[i];
                }
            }

            return max;
        }
    }
}

File: Solutions/C++/Arrays/MaximumSubarray.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int maxLength = nums[0], current = 0;

        for(int i = 0; i < nums.size(); i++) {
            if(current == 0 && nums[i] < 0)
                continue;

            current += nums[i];
            maxLength = (maxLength >= current ? maxLength : current);
            if(current < 0)
                current = 0;
        }

        return maxLength;
    }
};